lib/src/api_analysis/shapes_ext.dart

```diff
@@ -0,0 +1,228 @@
+  TypedefShape replaceWithNewIds({
+    required int newId,
+    required int? newTargetId,
+  }) =>
+      TypedefShape(
+        id: newId,
+        name: name,
+        targetClassId: newTargetId,
+      );
+}
+
+extension ExtensionShapeExt on ExtensionShape {
+  /// Creates a new [ExtensionShape] based on this one, but with the id and extendedClassId replaced.
+  ExtensionShape replaceWithNewIds({
+    required int newId,
+    required int newExtendedClassId,
+  }) =>
+      ExtensionShape(
+        id: newId,
+        name: name,
+        extendedClassId: newExtendedClassId,
+        getters: getters,
+        setters: setters,
+        methods: methods,
+        staticGetters: staticGetters,
+        staticSetters: staticSetters,
+        staticMethods: staticMethods,
+        annotations: annotations,
+      );
+}
+
+extension ClassShapeExt on ClassShape {
+  /// Creates a new [ClassShape] based on this one, but with the id replaced.
+  ClassShape replaceWithNewId(int newId) => ClassShape(
+        id: newId,
+        name: name,
+        getters: getters,
+        setters: setters,
+        methods: methods,
+        staticGetters: staticGetters,
+        staticSetters: staticSetters,
+        staticMethods: staticMethods,
+        unnamedConstructor: unnamedConstructor,
+        namedConstructors: namedConstructors,
+        annotations: annotations,
+      );
+
+  /// Modifies this [ClassShape] based on a list of others. The `name`, `id`,
+  /// `unnamedConstructor`, `namedConstructors` and `annotations` of the
+  /// original [ClassShape] are retained (subclasses don’t inherit constructors
+  /// or annotations from their superclass), and the other fields of type
+  /// [List<T>] are expanded with the corresponding fields from the elements of
+  /// [others], only where the addition of elements of the field [List]s does
+  /// not cause result in duplicate class members.
+  void add({required List<ClassShape> others}) {
+    /// Adds elements of [other] to [list] without creating duplicates.
+    void addWithoutDuplicates(
+      List<ClassMemberShapeBase> list,
+      List<ClassMemberShapeBase> other,
+    ) {
+      list.addAll(other.where((otherMember) => !list
+          .any((existingMember) => existingMember.name == otherMember.name)));
+    }
+
+    for (final otherClass in others) {
+      addWithoutDuplicates(getters, otherClass.getters);
+      addWithoutDuplicates(setters, otherClass.setters);
+      addWithoutDuplicates(methods, otherClass.methods);
+      addWithoutDuplicates(staticGetters, otherClass.staticGetters);
+      addWithoutDuplicates(staticSetters, otherClass.staticSetters);
+      addWithoutDuplicates(staticMethods, otherClass.staticMethods);
+    }
+  }
+}
+
+extension LibraryShapeExt on LibraryShape {
+  /// Creates a new [LibraryShape] based on this one, but with all the parts
+  /// (except `uri`) replaced.
+  LibraryShape replaceAll({
```

Comment:
these are extension methods so that the code in `normalizePackageShape()` is less cluttered
I made any extensions which are only used for normalization private and moved them into [normalize_shape.dart](https://github.com/dart-lang/pana/blob/a86cdd444191bbd3a1581b0d4927209bd157f78c/lib/src/api_analysis/normalize_shape.dart)